[Logs Explorer] Add customisation for virtual columns and add the 1st virtual column

achyutjhunjhunwala opened the pull request:
## Summary

- Closes https://github.com/elastic/kibana/issues/171844
- Closes https://github.com/elastic/kibana/issues/171726
- Closes  https://github.com/elastic/kibana/issues/173825

This PR does 4 things

- Customisation point to inject customCellRenderers in the Data Table.
- Adding the 1st virtual column - Content and removal of Message column
- Add cell actions for virtual columns so that the custom fields can be filtered for
- Renders similar content on the Flyout instead of Message

### To make this PR easy to review

#### Why Kibana Operations

- This PR initially increased the bundle size causing the values to go beyond the boundaries set in `limits.yml`. This has been fixed now after suggestion from Kibana Ops team

#### For Kibana Data Discovery Team - Changes done on Unified Data Table

- Moving around types to make them more generic and reusable
- Split the 2 components SourceDocument and SourceDocumentPopover out of the 1 single large file `ger_render_cell_value.tsx` they were in. This allows me to export them and lazy load the components in my plugin.
- Passed additional parameters to `externalCustomRenderers` function

#### For Kibana Data Discovery Team - Changes done on Discover Plugin

- Currently the discover plugin was not passing the registered customisation for `externalCustomRenderers`. Simply added the hook to retrieve the registered customisation and pass it as props.
- Added new customisation type for `CellRendererCustomisation`

#### For Logs UX Team - Changes done on Log Explorer

- Registered a new customisation for custom Virtual columns.
- Rendering logic for Log Level and Message with fallbacks
- When Message and its fallbacks are not present, we must render the whole document. Used extracted logic to display complete document as it appears in current Discover.
- When a JSON column is expanded (expandCellAction), it should render a JSON viewer as it currently does in Discover
- Items like Log Level which are rendered inside a Chip should allow contextual actions like Filter In, Filter Out and Copy

## Want to run this PR locally

```
// Checkout this PR
gh pr checkout 173732

// Start Kibana locally
yarn start 

// Start ES locally
yarn es snapshot

// Generate dummy data
node scripts/synthtrace simple_logs.ts --clean

```

## Live Demo can be seen [here](https://achyutjhunjhunwala-d-pr173732.kb.us-west2.gcp.elastic-cloud.com/app/observability-log-explorer/?pageState=(breakdownField:log.level,columns:!((field:service.name,width:240),(field:host.name,width:320),(field:content)),controls:(data_stream.namespace:(mode:include,selection:(selectedOptions:!(),type:options))),datasetSelection:(selectionType:all),filters:!(),query:(language:kuery,query:%27%27),refreshInterval:(pause:!t,value:5000),rowHeight:0,rowsPerPage:100,time:(from:%272024-01-05T14:39:41.024Z%27,to:%272024-01-05T14:54:43.378Z%27),v:1))

## Demo

![Virtual Column - Content](https://github.com/elastic/kibana/assets/7416358/db5977cd-1342-4dce-bfe9-bad3ee42487b)

## What's pending - Basically why is the PR still in Draft Stage

- [x] Fix colors for logs levels after confirmation from Isa and hide level when not present
- [x] Fix JSON rendering on the content cell when message field is not present so that it looks like exactly how it looks in Discover
- [x] Add E2E tests
- [x] Update DEMO with a gif rather than a screenshot

achyutjhunjhunwala committed:
Add customisation for virtual columns and add the 1st virtual column

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

achyutjhunjhunwala committed:
Revert changes to tests

achyutjhunjhunwala commented:
/ci

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Fix log level colors and visibility

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Add JSON formatting for content cells

achyutjhunjhunwala commented:
/ci

achyutjhunjhunwala committed:
Fix checktype issue

achyutjhunjhunwala committed:
Fix Logs explorer failing tests

achyutjhunjhunwala committed:
Fix snapshots for data-table

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala commented:
/ci

achyutjhunjhunwala committed:
Fix more tests

achyutjhunjhunwala committed:
Fix type issue

achyutjhunjhunwala committed:
Fix limits for the package

achyutjhunjhunwala commented:
/ci

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Add serverless tests

achyutjhunjhunwala committed:
Add stateful tests

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

elasticmachine commented:
Pinging @elastic/obs-ux-infra_services-team (Team:obs-ux-infra_services)

achyutjhunjhunwala commented on packages/kbn-unified-data-table/src/utils/get_render_cell_value.tsx:
```diff
@@ -109,6 +91,26 @@ export const getRenderCellValueFn = ({
       return <span className={CELL_CLASS}>-</span>;
     }
 
+    if (!!externalCustomRenderers && !!externalCustomRenderers[columnId]) {
```
2 changes were done here
1. Passed extra params to custom renderer
2. Moved the if condition after all the hooks have been called. This was anti-pattern but since its a function and not a React component, the linting is not catching this.

achyutjhunjhunwala replied:
2 changes were done here

Passed extra params to custom renderer
Moved the if condition after all the hooks have been called. This was anti-pattern but since its a function and not a React component, the linting is not catching this.

achyutjhunjhunwala commented on packages/kbn-unified-data-table/src/utils/get_render_cell_value.tsx:
```diff
@@ -133,36 +135,17 @@ export const getRenderCellValueFn = ({
-            );
-          })}
-        </EuiDescriptionList>
+        <SourceDocument
```
Extracted to make it reusable

achyutjhunjhunwala commented on packages/kbn-unified-data-table/src/utils/get_render_cell_value.tsx:
```diff
@@ -232,30 +270,22 @@ function renderPopoverContent({
-          />
-        </EuiFlexItem>
-      </EuiFlexGroup>
+      <SourcePopoverContent
```
Extracted to make it reusable

jbudz commented on packages/kbn-optimizer/limits.yml:
```diff
@@ -91,7 +91,7 @@ pageLoadAssetSize:
   links: 44490
   lists: 22900
-  logExplorer: 54342
+  logExplorer: 102152
```
Are we able to load this asynchronously?

achyutjhunjhunwala replied:
I will check this.

achyutjhunjhunwala replied:
@jbudz Thank you for the suggestion. I changed the logic to lazy load dependent components and now the bundle size is smaller

cauemarcondes reviewed (approved):
Obs UX LGTM

achyutjhunjhunwala committed:
Fix tests and added dynamic import to make the bundle smaller

achyutjhunjhunwala committed:
Fix silly mistake

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Fixing more type issues

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

kertal reviewed (commented):
Nice work making this part of the code more generic, customizable 👍. I tested and have some questions.
* the log level buttons have an arrow, this leads to the impression, they can trigger actions.  They don't and it would high likely cause a problem with the way EuiDataGrid handles focus of elements. So the question is, are these arrows necessary? Is it intended to trigger some actions further down the road
* Using a column based rendering could cause some issues, like the one in my screenshot
* Was it considered to use the log level as a separate column and use a field formatter for styling?
![Bildschirmfoto 2023-12-29 um 14 28 13](https://github.com/elastic/kibana/assets/463851/71698bb4-91ce-498e-8a27-06bffa8901fe)

achyutjhunjhunwala commented:
@kertal 
> the log level buttons have an arrow, this leads to the impression, they can trigger actions.

Yes, it will trigger similar actions, like the other cells do. We cannot use the Cell Actions to register actions and need them like this because, soon we will have multiple such cells. I am attaching the [Figma URL](https://www.figma.com/file/g2uboVjdzylHDyrIUrI4cU/%5BLogs%2B%5D-Virtual-columns?type=design&node-id=3277-317950&mode=design&t=g1lmZt0kTYQKiWBs-0) here for reference and a snapshot

<img width="789" alt="image" src="https://github.com/elastic/kibana/assets/7416358/52772f66-1f69-42c9-9ce1-c94988b5d405">


<img width="898" alt="image" src="https://github.com/elastic/kibana/assets/7416358/9f87ba9f-e18d-4711-9fbf-c801b87f68f4">

kertal commented:
> Yes, it will trigger similar actions, like the other cells do. I am working on adding custom cell actions for these virtual columns, soon we will have multiple such cells.

Thx, I this won't be added in this PR, it could make sense to remove the arrows, and add them when actions are available.

achyutjhunjhunwala committed:
Add logic to handle Cell Actions

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

achyutjhunjhunwala commented:
> Thx, I this won't be added in this PR, it could make sense to remove the arrows, and add them when actions are available.

@kertal I have added the actions now, I had to add them as part of this PR as without the cell actions, it would not make sense to release this virtual column as users won't be able to filter values

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/notice'

achyutjhunjhunwala committed:
Fix the stupidest and the most time consuming error

achyutjhunjhunwala committed:
Fix string literal issue

achyutjhunjhunwala committed:
Add cell action tests

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Fix flyout tests broken due to change in data-test-subj

jughosta reviewed (commented):
Hi @achyutjhunjhunwala,

Thanks for the changes/improvements in UnifiedDataTable! I left some suggestions in comments.

achyutjhunjhunwala commented on packages/kbn-unified-data-table/src/utils/get_render_cell_value.test.tsx:
```diff
@@ -219,64 +219,189 @@ describe('Unified data table cell rendering', function () {
-            Object {
-              "__html": 100,
-            }
+      <SourceDocument
```
Good idea, i would change them to check if `SourceDocuments` were used with correct props

jughosta replied:
Would it be possible to also add jest tests for the extracted components? To make sure that the inner content is still rendered as expected.

achyutjhunjhunwala committed:
Fix review comments from DD Team

achyutjhunjhunwala commented on packages/kbn-unified-data-table/src/utils/get_render_cell_value.test.tsx:
```diff
@@ -302,57 +427,46 @@ describe('Unified data table cell rendering', function () {
-              }
-            }
-            width={370}
+      <SourcePopoverContent
```
This SourcePopoverContent snapshots are not that verbose. So i left them there in the file. Let me know if you really want them to be changed as well

achyutjhunjhunwala replied:
discussed offline

weltenwort commented on x-pack/plugins/log_explorer/public/components/common/popover_chip.tsx:
```diff
@@ -0,0 +1,145 @@
+        <EuiFlexItem>
+          <div style={{ maxWidth: '200px' }}>
+            <EuiText size="s">
+              <span className="unifiedDataTable__cellPopoverValue eui-textBreakWord">
```
Referring to this "foreign" class name here seems brittle. Is there anything we can do to prevent accidental breakage? Maybe emulating the styles ourselves would be worth it?

achyutjhunjhunwala replied:
i placed myself in the shoes of the Unified Data Table developer who lets say introduces a change.

- If they change the CSS itself, then our will automatically be updated as we want uniform experience.
- If they decide to change the className itself, i would ideally do a findAll to see where it has been used and update accordingly.

I know the above does not cover all the edge cases, let's say they decide to add an additional class to the Popover content to get some additional style, then our won't get it. But in that case even if we copy paste the CSS, that would still mean the same.

Also as you mentioned in a later comment, we might move away from this component to UiActions, in that case this code is not going to live for long.

But yeah this is a brittle piece. Safest option is to style it ourselves. I will copy over the CSS 👍🏼

jughosta replied:
There is also an option to introduce a wrapper component in UnifiedDataTable which includes these styles. Then you could import it in your code safely.

weltenwort commented on x-pack/plugins/log_explorer/public/components/common/popover_chip.tsx:
```diff
@@ -0,0 +1,145 @@
+          />
+        </EuiFlexItem>
+      </EuiFlexGroup>
+      <EuiPopoverFooter>
```
Just a thought: At some point it might be worth re-implementing this based on the UiActions mechanism that allows for registering additional triggers. But this is probably fine for the initial version.

weltenwort commented on x-pack/plugins/log_explorer/public/components/common/translations.ts:
```diff
@@ -122,22 +122,30 @@ export const flyoutShipperLabel = i18n.translate('xpack.logExplorer.flyoutDetail
 });
 
-export const flyoutHoverActionFilterForText = (text: unknown) =>
+export const actionFilterForText = (text: unknown) =>
```
Can you help me understand why the `text` is typed as `unknown` here instead of `string`?

achyutjhunjhunwala replied:
Thank you for bringing this to attention. Initially this was not properly typed due to us directly using the type from `DataTableRecord['flattened']` which has a type of `flattened: Record<string, unknown>;`

https://github.com/elastic/kibana/blob/4f9e2bdf195caace9ce860230929c86e31cacb6a/packages/kbn-discover-utils/src/types.ts#L20-L37

This should not be the case anymore. I will fix it 👍🏼

weltenwort commented on x-pack/plugins/log_explorer/public/components/flyout_detail/sub_components/log_level.tsx:
```diff
@@ -6,27 +6,55 @@
 }
 
-export function LogLevel({ level }: LogLevelProps) {
+export function LogLevel({
```
Would it make sense for this file to be in `components/common` now?

weltenwort commented on x-pack/plugins/log_explorer/public/components/flyout_detail/sub_components/log_level.tsx:
```diff
@@ -6,27 +6,55 @@
+    ? euiTheme.colors[LEVEL_DICT[level as keyof typeof LEVEL_DICT]]
+    : null;
+
+  if (renderInFlyout) {
```
Given the two branches are almost equal in size, would it make sense to create two individual components instead of one with a boolean flag?

achyutjhunjhunwala replied:
I initially built 2 separate components, `log_level_flyout` and `log_level_table`. I found both them sharing almost the same code with only 2 difference.

1. Flyout Log Level will not have icons
2. Flyout Log level should not be a clickable item as it has its own popover on hover. 

Given these 2 challenges can be prop controlled i chose to branch them out. But lmk, if you still think we should split them

weltenwort reviewed (commented):
This is the first part of my review, which I already wanted to share so I don't block you for too long. I found it to be a really good read so far and what I saw was really well done - especially given the complexity of the feature. :clap:

I left a few questions in-line. Additionally I observed the following:

**DataView warning toast:** When adding a filter from the log level chip I saw a warning toast about data views not matching up:

![Pasted image 20240103174803](https://github.com/elastic/kibana/assets/973741/a524e3aa-fb97-4fa3-9c31-31664ae41d16)

The filter works, but there seems to be something in the filter manager that wants it to be associated with the data view. Maybe it's something we need to pass?

**Message prefix:** Currently the field name prefix is always shown, even when the expected case of showing the message is met:

![image](https://github.com/elastic/kibana/assets/973741/042a00fb-b335-488c-928b-62480b94b094)

Do we want to save some screen estate by hiding it if not falling back to any of the other fields? Or is the inconsistency too big?

**Content in fly-out:** If I read the ACs correctly the fall-back logic should also be applied to the message display in the fly-out. That doesn't seem to be the case right now. Only the log level component is re-used.

achyutjhunjhunwala commented:
@weltenwort Thank you so much for the initial review. Answering the top level comment here and the rest inline

> DataView warning toast: When adding a filter from the log level chip I saw a warning toast about data views not matching up:

I thought its my local which is doing something wrong. I will have a look at it

> Message prefix: Currently the field name prefix is always shown, even when the expected case of showing the message is met:

I added the message prefix as when none of the message fields are present, we will render the the whole document. In that case we are not adding Prefix for the Document. So not to confuse between the 2 fields i did it like that

message (or fallback) is Present -> prefix+message
Not present -> Only document

> Content in fly-out:

Good catch on that, completely missed it. Will add it now

achyutjhunjhunwala commented on x-pack/plugins/log_explorer/public/components/common/copy_button.tsx:
```diff
@@ -0,0 +1,31 @@
+        iconType="copyClipboard"
+        aria-label={ariaCopyValueText}
+        onClick={() => copyToClipboard(value as string)}
+        data-test-subj={`dataTableCellAction_copyToClipboardAction_${property}_${value.replace(
```
True, i will remove the value prop.

achyutjhunjhunwala replied:
I have removed the `value`

achyutjhunjhunwala commented on src/plugins/discover/public/customizations/customization_types/cell_renderer_customisation.ts:
```diff
@@ -0,0 +1,14 @@
+import type { CustomCellRenderer } from '@kbn/unified-data-table';
+
+export interface CellRendererCustomisation {
+  id: 'cell_renderer';
```
I initially thought of naming this `data_table_column` but then that would have been a very broad scoped name given we actually only control the `cell rendering`. For example, in future we will build another customisation for `column header tooltips` and may be more around `cell actions`.

So to narrow the scope, i chose to name is `cell_renderer`.

achyutjhunjhunwala replied:
I see your point now. 👍🏼

achyutjhunjhunwala commented on x-pack/plugins/log_explorer/public/components/common/filter_in_button.tsx:
```diff
@@ -0,0 +1,49 @@
+import {
+  createFilter,
+  isEmptyFilterValue,
+} from '@kbn/cell-actions/src/actions/filter/create_filter';
```
In order to fix that DataView issue, i used a different approach and now we don't need this
I used this
`import { generateFilters } from '@kbn/data-plugin/public';`

weltenwort reviewed (commented):
> I added the message prefix as when none of the message fields are present, we will render the the whole document. In that case we are not adding Prefix for the Document. So not to confuse between the 2 cases i added prefix for message as well.
> 
> message (or fallback) is Present -> prefix+message
> Not present -> Only document

But if we render the full document, each entry will consist of key-value pairs. So there should be no ambiguity even if we remove the message prefix, right?

Maybe this would be a good point in time to ask @isaclfreire to also review from a design perspective. What do you think?

achyutjhunjhunwala commented on x-pack/plugins/log_explorer/common/constants.ts:
```diff
@@ -27,12 +29,18 @@ export const LOG_FILE_PATH_FIELD = 'log.file.path';
 export const DEFAULT_COLUMNS = [
+  {
+    field: CONTENT_FIELD,
+  },
```
TBH, no rationale. So timestamp is always the 1st column. As per the new designs, right after the timestamp we will place Resource Column which is not yet there and then Content Column. So i placed Content as 1st after timestamp. Once we add Resource, that would actually take this spot and Content would move next

weltenwort replied:
Right... I stumbled across this, because the current host name and service name columns are sort-of the predecessor for the resource column. So what if we put them in the same location that the resource column will be in in the future?

achyutjhunjhunwala replied:
Sounds good

achyutjhunjhunwala commented on x-pack/plugins/log_explorer/public/components/common/popover_chip.tsx:
```diff
@@ -0,0 +1,145 @@
+        font-size: ${xsFontSize};
+        display: flex;
+        justify-content: center;
+        ${style ? css(convertStyleToString(style)) : ''}
```
I was actually not sure over the preference if `css` gets rendered first and then `style` is applied or the other way. 
But now i tested it and yes, style overrides css so all good. Code changed 👍🏼

weltenwort replied:
Yes, if I understand `emotion` correctly, `css` gets turned into `class`. So unless it pulls any `important` tricks its styles should have lower specificity than `style`.

achyutjhunjhunwala committed:
Fix review comments except 1 around Flyout Message

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

achyutjhunjhunwala committed:
Set content column at the end

achyutjhunjhunwala committed:
Fix flyout message rendering logic

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

weltenwort reviewed (commented):
The data view warning is gone now, well done. We're almost there, I think.

One small thing I noticed is, that we don't have the ability to filter-in/filter-out the message or its fallsbacks via the cell pop-over.

![image](https://github.com/elastic/kibana/assets/973741/f236c71c-d141-42ca-a282-50ada2e4e820)

Do you know if that would be possible to add back? It works in the fly-out content section:

![image](https://github.com/elastic/kibana/assets/973741/565029da-fc37-4555-a665-8ac8b208147d)

achyutjhunjhunwala commented on x-pack/plugins/log_explorer/public/components/common/popover_chip.tsx:
```diff
@@ -112,7 +110,9 @@ export function ChipWithPopover({
             <EuiText size="s">
-              <span className="unifiedDataTable__cellPopoverValue eui-textBreakWord">{text}</span>
+              <DataTableCellValue>
+                {property}:{text}
```
@weltenwort We started the discussion here - https://www.figma.com/file/g2uboVjdzylHDyrIUrI4cU?node-id=3277:320484&mode=design#658525245

achyutjhunjhunwala replied:
discussed offline, new design would mimic the Source Column style i.e., `<bold>{property}</bold> {text}`

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

jughosta commented on packages/kbn-unified-data-table/src/components/source_document.tsx:
```diff
@@ -0,0 +1,127 @@
+
+const CELL_CLASS = 'unifiedDataTable__cellValue';
+
+export function SourceDocument({
```
Could you please add a snapshot test? It's an important part of Discover and it would be great not to reduce the coverage by the refactoring.

jughosta commented on packages/kbn-unified-data-table/src/utils/get_render_cell_value.tsx:
```diff
@@ -6,31 +6,24 @@
 import { UnifiedDataTableContext } from '../table_context';
-import { defaultMonacoEditorWidth } from '../constants';
-import JsonCodeEditor from '../components/json_code_editor/json_code_editor';
+import { CustomCellRenderer } from '../..';
```
Would be better to import from the types file only so it does not depend on the other included exports.

achyutjhunjhunwala replied:
If i understand correctly, you mean explicit from the file name -
```import type { CustomCellRenderer } from '../types';```

jughosta commented on src/plugins/discover/public/customizations/customization_types/data_table_customisation.ts:
```diff
@@ -0,0 +1,14 @@
+
+export interface DataTableCustomisation {
+  id: 'data_table';
+  customCellRenderer: CustomCellRenderer;
```
```suggestion
  customCellRenderer?: CustomCellRenderer;
```

jughosta reviewed (commented):
Thanks for addressing the previous comments!

gbamparop commented:
Is the `@timestamp` or another field supposed to be bold when the whole JSON document is rendered in the content column?

<img width="1235" alt="image" src="https://github.com/elastic/kibana/assets/5831975/01aba52a-ea90-43f4-aab0-6a9a69f42eb9">

achyutjhunjhunwala commented:
> Is the `@timestamp` or another field supposed to be bold when the whole JSON document is rendered in the content column?

@gbamparop Thats the default Discover behaviour. It marks all key value property like that. All keys are styled bold. You can test this by going into Discover and remove all columns. Then Discover would only render Timestamp column and the Whole JSON Document column

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala commented:
> The data view warning is gone now, well done. We're almost there, I think.
> 
> One small thing I noticed is, that we don't have the ability to filter-in/filter-out the message or its fallsbacks via the cell pop-over.
> 
> ![image](https://private-user-images.githubusercontent.com/973741/294247695-f236c71c-d141-42ca-a282-50ada2e4e820.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0NjM1NTUsIm5iZiI6MTcwNDQ2MzI1NSwicGF0aCI6Ii85NzM3NDEvMjk0MjQ3Njk1LWYyMzZjNzFjLWQxNDEtNDJjYS1hMjgyLTUwYWRhMmU0ZTgyMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwNVQxNDAwNTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MzVhNjRiZGI3NDY4MTNjM2IxMzFjMjYxNDc4NDYzYjE1YmQzNTRmYmMwMTYzOTY5NjljNDFmMWNjNzdjMjEzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.YLrCVKrdkHITCvx-XGDGJdgCZM_Snb8eGZ6pE7Vpmn0)
> 
> Do you know if that would be possible to add back? It works in the fly-out content section:
> 
> ![image](https://private-user-images.githubusercontent.com/973741/294247929-565029da-fc37-4555-a665-8ac8b208147d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ0NjM1NTUsIm5iZiI6MTcwNDQ2MzI1NSwicGF0aCI6Ii85NzM3NDEvMjk0MjQ3OTI5LTU2NTAyOWRhLWZjMzctNDU1NS1hNjY1LThhYzhiMjA4MTQ3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwNVQxNDAwNTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzdiM2MwMjFmYjY5OWZiZjI3NzFmZWI2MTM4MzNjOGJmY2I1OTZkZDg1ZDZmZGZhOWFjZGRjMzhkY2E3MzMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ufSB0NoJOiuKs1v7Maf1Vnecko-Rq1-SKGjZw23rlvY)

@weltenwort @isaclfreire 

I am not how the filter action on the Table would look like for the message field.
It brings certain complexity which are not present in the Flyout.

1. In the Flyout, we only render the Content block when message or any fallbacks are present. Where as in the table we render the Raw Documents when none is present. In this case we don't want user to add that as a filter. This means not all cells in the Table would have let say the same actions. Not sure if we want that.

Also the Log Level Chip provides a Arrow on the right hand side giving visual hints that there is a possible action if clicked on. For Message field, we don't have any such visual hint. I am not sure how we want to handle this. Open for suggestions

achyutjhunjhunwala committed:
Fix further feedback comments

weltenwort commented:
Visually these just would be the data grid hover and popover actions. For what happens when the user clicks two options come to mind:

- Just filter for the `message`/`error.message`/`event.original`. This would mean there's no action for the "full document" fall-back and this wouldn't work for the resource column, because there's no obvious "main" field.
- Filter for a combination of the log level and message. This still wouldn't make sense for the "full document", but at least it would be applicable for the resource column (i.e. "filter for this exact combination").

We can split this change into a follow-up PR to get this one merged soon if you like.

achyutjhunjhunwala commented:
@weltenwort @isaclfreire Given the size, complexity of the PR and next week being On-Week, i am in favour of splitting the task of how we handle Filtering behavior for Message Field in the Table separately out side the scope of this PR

achyutjhunjhunwala committed:
Fix broken Unit tests

weltenwort commented:
Ok, let's tackle the cell actions in a follow-up. Even if we don't want to support the filter actions, the "copy" action would probably be quite useful for users.

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

davismcphee reviewed (commented):
Great work! I left a few minor points of feedback, but assuming those get addressed, the Discover changes LGTM.

I didn't do extensive testing on the Logs Explorer side, so my review is primarily on the Discover changes, but I poked around a bit and noticed a couple things worth mentioning. I'll approve once the Discover feedback has been addressed though, and leave it to your team to decide what to do with this feedback.

It looks like removing a column from the sidebar causes Logs Explorer to crash with a React hooks related error:

https://github.com/elastic/kibana/assets/25592674/d335e1c2-3048-49b5-b702-4a467a78b9aa

After opening a log level badge popover, clicking on the badge again to close it causes the popover to get stuck open. This is most often caused by using `onClick={() => setIsPopoverOpen(true)}` instead of `onClick={() => setIsPopoverOpen(!isPopoverOpen)}`:

https://github.com/elastic/kibana/assets/25592674/9e5f45f8-8c82-42e2-82de-de1383a8afd1

These next ones aren't necessarily bugs, just some UX feedback.

Unfortunately it looks like the EUI upgrade broke the vertical alignment on the log level badge:
<img width="1266" alt="alignment" src="https://github.com/elastic/kibana/assets/25592674/994d275e-24b1-46be-851f-46425a136b15">

I agree with @kertal about the column based approach to the content cell. For example, it causes the cell layout to look odd IMO when setting the row height to auto:
<img width="974" alt="column" src="https://github.com/elastic/kibana/assets/25592674/52a613c0-3b43-4210-8f96-edd2113618f2">

It also feels like by default there's too little space available for the content cell compared to the service.name and host.name cells. This can be improved by collapsing the sidebars, but it might be worth reconsidering the defaults here IMO:
<img width="1995" alt="content" src="https://github.com/elastic/kibana/assets/25592674/dc3ee37f-6542-42e7-9edc-f38a70f32270">

achyutjhunjhunwala commented on x-pack/plugins/log_explorer/public/components/virtual_columns/content.tsx:
```diff
@@ -135,9 +130,15 @@ const Content = ({
+  ({ data }: { data: LogExplorerDiscoverServices['data'] }) =>
+  (props: DataGridCellValueElementProps) => {
+    const { dataView } = props;
+    const virtualColumnServices = {
```
@weltenwort Somehow adding useMemo here is breaking the component.
Check 1st video from Davis's comment. I am removing it for the time being

weltenwort replied:
ok, so this function is not actually a component... good to know :wink:

achyutjhunjhunwala committed:
Fix discover teams review feedback

achyutjhunjhunwala commented:
@davismcphee Thank you for the detailed review. Looks like i introduced some regressions while fixing other code reviews. I have fixed most of them, answering them here,

> It looks like removing a column from the sidebar causes Logs Explorer to crash with a React hooks related error:

Good catch, fixed it now. 

> After opening a log level badge popover, clicking on the badge again to close it causes the popover to get stuck open

Fixed

> Unfortunately it looks like the EUI upgrade broke the vertical alignment on the log level badge:

I didn't understand this one. I see this as expected. The text and the arrow is center aligned to the badge. Which means they will not be vertically aligned with each other.

> I agree with @kertal about the column based approach to the content cell. For example, it causes the cell layout to look odd IMO when setting the row height to auto:

I am not sure if we are setting row height as Auto in Log Explorer for the data table. Or is it some user level setting ?
I will loop in @weltenwort and @isaclfreire to comment on here.

> It also feels like by default there's too little space available for the content cell compared to the service.name and host.name cells. This can be improved by collapsing the sidebars, but it might be worth reconsidering the defaults here IMO:

We have more [user stories coming](https://github.com/elastic/kibana/pull/173732#issuecomment-1872104340) in where the regular columns like `service.name` and `host.name` will be replaced by a single column called `resource` column. This should create more space for the content column. 
@isaclfreire fyi

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

achyutjhunjhunwala committed:
Fixed design issues post discussion

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column

davismcphee reviewed (approved):
Thanks for addressing the PR feedback, the Data Discovery changes LGTM 👍 And I confirmed the sidebar and popover issues are working as expected now too.

>I didn't understand this one. I see this as expected. The text and the arrow is center aligned to the badge. Which means they will not be vertically aligned with each other.

Sorry, I should have been a bit more specific with what I meant here. It looks like originally the badge was vertically aligned to the center of the row, but after the last EUI upgrade it's aligned to the bottom of the row instead.

Before upgrade:
<img width="574" alt="aligned" src="https://github.com/elastic/kibana/assets/25592674/2233f1e8-dff3-4cd1-b0f7-661ccd8c6539">

After upgrade:
<img width="1266" alt="misaligned" src="https://github.com/elastic/kibana/assets/25592674/53ceb319-934b-4361-aad8-fce64bf699aa">

>I am not sure if we are setting row height as Auto in Log Explorer for the data table. Or is it some user level setting ?

By default it seems to be set to single, but this can be changed by users through the display options button above the grid:
<img width="503" alt="display_options" src="https://github.com/elastic/kibana/assets/25592674/5a762a8d-fdee-4dbc-b50b-fcc58d0d84d0">

>We have more user stories coming in where the regular columns like `service.name` and `host.name` will be replaced by a single column called `resource` column. This should create more space for the content column.

This makes sense, and I agree it would likely improve the current situation with limited space.

>We would want to keep the Log Level as part of the Content column for the moment. The table would already limit the number of columns when it loads initially to provide enough space for the Content column.

Oops, my feedback here was also a bit vague. I didn't mean that I think a separate data table column would be better for the log level, just that the current use of flex columns  results in a lot of whitespace when the content wraps. Placing the badge inline with the content so the text can wrap around it might make better use of the available space.

That said, none of this feedback is blocking or anything, so no need for me to hold up this PR 🙂

achyutjhunjhunwala commented:
@davismcphee Thank you for the approval
1. Regarding the EUI change, i discussed this with the EUI team on the Slack channel. Post that we decided to leave it like that for the moment.
2. Same with the multiline change via the control. At the moment we default it to single, but i agree users can change it. i discussed this with our team internally and we will reiterate this over time to improve it.

weltenwort reviewed (approved):
Great job! :clap: Thanks for all the attention to detail!

achyutjhunjhunwala committed:
Merge branch 'main' into add-content-virtual-column
